pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,569 @@
+	}
+	// Remove endpoints missing from the update.
+	for service := range proxier.serviceMap {
+		if _, exists := registeredEndpoints[service]; !exists {
+			glog.V(2).Infof("Removing endpoints for %s", service)
+			proxier.serviceMap[service].endpoints = nil
+		}
+	}
+}
+
+// used in OnEndpointsUpdate
+type hostPortPair struct {
+	host string
+	port int
+}
+
+func isValidEndpoint(hpp *hostPortPair) bool {
+	return hpp.host != "" && hpp.port > 0
+}
+
+// Tests whether two slices are equivalent.  This sorts both slices in-place.
+func slicesEquiv(lhs, rhs []string) bool {
+	if len(lhs) != len(rhs) {
+		return false
+	}
+	if reflect.DeepEqual(slice.SortStrings(lhs), slice.SortStrings(rhs)) {
+		return true
+	}
+	return false
+}
+
+func flattenValidEndpoints(endpoints []hostPortPair) []string {
+	// Convert Endpoint objects into strings for easier use later.
+	var result []string
+	for i := range endpoints {
+		hpp := &endpoints[i]
+		if isValidEndpoint(hpp) {
+			result = append(result, net.JoinHostPort(hpp.host, strconv.Itoa(hpp.port)))
+		}
+	}
+	return result
+}
+
+// servicePortToServiceChain takes the ServicePortName for a
+// service and returns the associated iptables chain
+// this is computed by hashing (sha256) then encoding to base64 and
+// truncating with the prefix "KUBE-SVC-"
+// We do this because Iptables Chain Names must be <= 28 chars long
+func servicePortToServiceChain(s proxy.ServicePortName) utiliptables.Chain {
+	hash := sha256.Sum256([]byte(s.String()))
+	encoded := base64.URLEncoding.EncodeToString(hash[:])
+	return utiliptables.Chain("KUBE-SVC-" + encoded[:19])
+}
+
+// this is the same as servicePortToServiceChain but with the endpoint included essentially
+func servicePortAndEndpointToServiceChain(s proxy.ServicePortName, endpoint string) utiliptables.Chain {
+	hash := sha256.Sum256([]byte(s.String() + "_" + endpoint))
+	encoded := base64.URLEncoding.EncodeToString(hash[:])
+	return utiliptables.Chain("KUBE-SVC-" + encoded[:19])
+}
+
+// This is where all of the iptables-save/restore calls happen.
+// The only other iptables rules are those that are setup in iptablesInit()
+// assumes proxier.mu is held
+func (proxier *Proxier) syncProxyRules() error {
+	glog.V(2).Infof("Syncing iptables rules.")
+
+	// ensure main chain and rule connecting to output
+	args := []string{"-m", "comment", "--comment", "handle services"}
+	if _, err := proxier.iptables.EnsureChain(utiliptables.TableNAT, iptablesServicesChain); err != nil {
+		return err
+	}
+	if _, err := proxier.iptables.EnsureRule(utiliptables.Prepend, utiliptables.TableNAT, utiliptables.ChainOutput, append(args, "-j", string(iptablesServicesChain))...); err != nil {
+		return err
+	}
+	if _, err := proxier.iptables.EnsureRule(utiliptables.Prepend, utiliptables.TableNAT, utiliptables.ChainPrerouting, append(args, "-j", string(iptablesServicesChain))...); err != nil {
+		return err
+	}
+
+	// Get iptables-save output so we can check for existing chains and chainsLines.
```

Comment:
whoops, i thought i undid all of those. yeah i initially had the brilliant
idea to do find and replace with a regex for the rulesLines and chainLines
rename and it made a bit of a mess. fixing next commit.

On Tue, Aug 11, 2015 at 3:59 PM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/proxy/iptables/proxier.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#discussion_r36791073>
> :
>
> > +func (proxier *Proxier) syncProxyRules() error {
> > +	glog.V(2).Infof("Syncing iptables rules.")
> > +
> > +	// ensure main chain and rule connecting to output
> > +	args := []string{"-m", "comment", "--comment", "handle services"}
> > +	if _, err := proxier.iptables.EnsureChain(utiliptables.TableNAT, iptablesServicesChain); err != nil {
> > +		return err
> > +	}
> > +	if _, err := proxier.iptables.EnsureRule(utiliptables.Prepend, utiliptables.TableNAT, utiliptables.ChainOutput, append(args, "-j", string(iptablesServicesChain))...); err != nil {
> > +		return err
> > +	}
> > +	if _, err := proxier.iptables.EnsureRule(utiliptables.Prepend, utiliptables.TableNAT, utiliptables.ChainPrerouting, append(args, "-j", string(iptablesServicesChain))...); err != nil {
> > +		return err
> > +	}
> > +
> > +	// Get iptables-save output so we can check for existing chains and chainsLines.
>
> I think you substituted "chainLines" for "rules" in this comment :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210/files#r36791073>
> .
>